File: cpp/2021_02/TargetSum.cc
/** 494. Target Sum */
/** You are given a list of non-negative integers, a1, a2, ..., an, and a target, S. 
 *  Now you have 2 symbols + and -. For each integer, you should choose one from + and - as its new symbol.
 *  Find out how many ways to assign symbols to make sum of integers equal to target S.
 * 
 *  Example 1:
 *  Input: nums is [1, 1, 1, 1, 1], S is 3. 
 *  Output: 5
 * 
 *  Explanation: 
 *  -1+1+1+1+1 = 3
 *  +1-1+1+1+1 = 3
 *  +1+1-1+1+1 = 3
 *  +1+1+1-1+1 = 3
 *  +1+1+1+1-1 = 3
 *  There are 5 ways to assign symbols to make the sum of nums be target 3.
 * 
 *  Constraints:
 *  The length of the given array is positive and will not exceed 20.
 *  The sum of elements in the given array will not exceed 1000.
 *  Your output answer is guaranteed to be fitted in a 32-bit integer. **/

// dfs는 내가 처음 생각한 풀이. 

#include <vector>
#include <cstring>
#include <iostream>

using namespace std;

int dp[1000];

int subsetSum(vector<int> &nums, int s) {
    for (int num : nums) {
        for (int i = s; i >= num; --i) {
            dp[i] += dp[i-num];
        }
    }

    return dp[s];
}

void dfs(vector<int> &nums, int index, int sum, int S, int &answer) {
    if (index == nums.size()) {
        if (sum == S) answer++;
        return;
    }

    dfs(nums, index+1, sum-nums[index], S, answer);
    dfs(nums, index+1, sum+nums[index], S, answer);
}

int findTargetSumWays(vector<int> &nums, int S) {
    int sum = 0;
    
    memset(dp, 0 , sizeof(dp));
    dp[0] = 1;
    for (int num : nums) sum += num;

    return sum < S || (sum + S) % 2 == 1 ? 0 : subsetSum(nums, (sum+S) / 2);

    //dfs(nums, 0, 0, S, answer);
    //return answer;
}

int main() {
    vector<int> nums0 {1,1,1,1,1};
    int ans0 = findTargetSumWays(nums0, 3);

    cout << ans0 << endl; // 5

    return 0;
}

File: python/src/year_twenty_two/november/target_sum.py
"""
494. Target Sum

You are given an integer array nums and an integer target.
You want to build an expression out of nums by adding one of the symbols '+' and '-' before
each integer in nums and then concatenate all the integers.
For example, if nums = [2, 1], you can add a '+' before 2 and a '-' before 1 and
concatenate them to build the expression "+2-1".
Return the number of different expressions that you can build, which evaluates to target.

Example 1:
Input: nums = [1,1,1,1,1], target = 3
Output: 5
Explanation: There are 5 ways to assign symbols to make the sum of nums be target 3.
-1 + 1 + 1 + 1 + 1 = 3
+1 - 1 + 1 + 1 + 1 = 3
+1 + 1 - 1 + 1 + 1 = 3
+1 + 1 + 1 - 1 + 1 = 3
+1 + 1 + 1 + 1 - 1 = 3

Example 2:
Input: nums = [1], target = 1
Output: 1

Constraints:
1 <= nums.length <= 20
0 <= nums[i] <= 1000
0 <= sum(nums[i]) <= 1000
-1000 <= target <= 1000

"""

from typing import List
from collections import Counter

MAX = 1000


class Solution:
    def __init__(self) -> None:
        self.cache = [[0 for i in range(20001)] for j in range(21)]

    def save(self, index: int, target: int, value: int):
        self.cache[index + 1][target + MAX] = value

    def load(self, index: int, target: int):
        return self.cache[index + 1][target + MAX]

    def find_target_sum_ways_dp(self, nums: List[int], target: int) -> int:
        self.save(-1, 0, 1)

        for i, num in enumerate(nums):
            subsum = sum(nums[: i + 1])

            for j in range(-subsum, subsum + 1):
                a = self.load(i - 1, j - num)
                b = self.load(i - 1, j + num)
                self.save(i, j, a + b)

        return self.load(len(nums) - 1, target)

    def find_target_sum_ways(self, nums: List[int], target: int) -> int:
        counter = Counter({0: 1})

        for num in nums:
            step = Counter()

            for prev_sum, count in counter.items():
                step[prev_sum + num] += count
                step[prev_sum - num] += count

            counter = step

        return counter[target]
